hoss/dimension_utilities.py

```diff
@@ -75,8 +75,143 @@ def prefetch_dimension_variables(opendap_url: str, varinfo: VarInfoFromDmr,
+
+        For each dimension variable:
+        (1) Check if the variable needs a bounds variable.
+        (2) If so, create a bounds array of minimum and maximum values.
+        (3) Then write the bounds variable to the NetCDF4 URL.
+
+    """
+    with Dataset(dimensions_nc4, 'r+') as datasets:
+        for dimension_name in required_dimensions:
+            dimension_variable = varinfo.get_variable(dimension_name)
+            if needs_bounds(dimension_variable) is True:
+                min_and_max_bounds = create_bounds(datasets, dimension_name)
+                write_bounds(datasets, dimension_variable, min_and_max_bounds)
+
+                logger.info('Artificial bounds added for dimension variable: '
+                f'{dimension_name}')
+
+        
+def needs_bounds(dimension: VariableFromDmr) -> bool:
+    """ Check if a dimension variable needs a bounds variable.
+        This will be the case when dimension cells are edge-aligned 
+        and bounds for that dimension do not already exist.
+    
+    """
+    
+    return dimension.attributes['cell_alignment'] == 'edge' and dimension.references.get('bounds') == None
+
+
+def create_bounds(dimension_dataset: Dataset,
+                  dimension_path: str) -> np.ndarray:
+    """ Create an array containing the minimum and maximum bounds
+        for a given dimension. 
+
+        The minimum and maximum values are determined under the assumption
+        that the dimension data is monotonically increasing and contiguous.
+        So for every bounds but the last, the bounds are simply extracted
+        from the dimension dataset.
+
+        The final bounds must be calculated with the assumption that 
+        the last data cell is edge-aligned and thus has a value the does
+        not account for the cell length. So, the final bound is determined
+        by taking the median of all the resolutions in the dataset to obtain
+        a resolution that can be added to the final data value. 
+
+        Ex: Input dataset with resolution of 3 degrees:  [ ... , 81, 84, 87]
+
+        Minimum | Maximum
+         <...>     <...>
+          81        84
+          84        87
+          87        ?  ->  87 + median resolution -> 87 + 3 -> 90
+        
+    """
+    # Access the dimension variable's data using the variable's full path.
+    dimension_data = dimension_dataset[dimension_path][:]
+
+    # Determine the dimension's resolution by taking the median value 
+    # of the differences between each ascending data value.
+    dimension_array = np.array(dimension_data)
+
+    # Build array.
+    size = dimension_array.size
+    min_max_pairs = [[dimension_array[idx], 
+                      dimension_array[idx+1]] 
+                      for idx in range(0, size-1)]
+
+    # Calculate final values.
+    dim_resolution = np.median(np.diff(dimension_array))
+    min_max_pairs.append([dimension_array[size-1], 
+                          dimension_array[size-1] + dim_resolution])
```

Comment:
I assume this is tested, so I'm suspicious of this observation, but...  On line 160 above the min_max_pairs are being populated for the range (0 .. size-1).  And then, here, an additional value is being populated.  Wouldn't that be one beyond what we want to populate?  Also on line 159, referencing dimension_array[idx+1] for idx = (size-1) - wouldn't that be one beyond the range of dimension_array?

Reply:
Oh right! python ranges only include up until the second value, but don't include that value itself. So the final idx value in the min_max_pairs array loop will be size-2, since range(0, size-1).

So if the size of the dimension array is, say, 60, the loop will write pair values up to index 58, then the final value calculation adds the [dimension_array[60-1], dimension_array[60-1] + res] which amounts to a final index of 59.